fetch_from_gmail.py

```diff
@@ -0,0 +1,374 @@
+            }
+            json.dump(
+                {
+                    "threads_dict": serializable_threads,
+                    "last_page_token": last_page_token,
+                },
+                f,
+            )
+        finally:
+            fcntl.flock(f, fcntl.LOCK_UN)
+    # Atomic rename operation
+    os.replace(temp_file, "fetching_state.json")
+
+
+def load_state_from_json_file():
+    global threads_dict, last_page_token, seen_email_ids
+    if not os.path.exists("fetching_state.json"):
+        threads_dict = {}
+        last_page_token = None
+        seen_email_ids = set()
+        return
+    with open("fetching_state.json", "r") as f:
+        try:
+            state = json.load(f)
+            # Convert serialized threads back to EmailThread objects
+            loaded_threads = state.get("threads_dict", {})
+            threads_dict = {
+                thread_id: EmailThread.from_dict(thread_data)
+                for thread_id, thread_data in loaded_threads.items()
+            }
+            last_page_token = state.get("last_page_token")
+            seen_email_ids = set()
+            for thread_id in threads_dict:
+                for email in threads_dict[thread_id].emails:
+                    seen_email_ids.add(email.id)
+        except Exception as e:
+            print(f"Error loading state from json file: {e}")
+            threads_dict = {}
+            last_page_token = None
+            seen_email_ids = set()
+
+
+def fetch_single_message(
+    creds,
+    msg_id,
+    seen_email_ids,
+    pagination_token_that_was_loaded_from_file,
+    last_page_token,
+):
+    """Fetch a single message and return (Email object, thread_id) tuple or special string values."""
+    if (
+        msg_id in seen_email_ids
+        and pagination_token_that_was_loaded_from_file != last_page_token
+    ):
+        # We have pagination_token_that_was_loaded_from_file != last_page_token condition because
+        # if we did not have it, then we may miss out on certain emails that were not saved in the current
+        # page from a previous run. For example, if in the previous run, pagination token was p1, and we saved
+        # 5 out of the 100 emails for that page, then in the next run, if we do not have this condition, we may
+        # quit before saving the remaining 95 emails. Now, having this condition means that we will not stop
+        # for the current page, as long as the pagination token is a different one from the one that was last
+        # used from the previous run. So this way, in the next run, after we paginate through all 100 emails of the
+        # current page, and then on the next page, we encounter an email that we have already seen, then we will
+        # can be sure that we have seen all emails in and before that page (assuming the BATCH SIZE has not been reduced compared to previous runs).
+        return "stop_early"  # Special return value to indicate continue
```

Comment:
This is a bit of a smell. Lets make this into a constant so that it is easier to check for than string comparison.

Reply:
I wonder if you could use exceptions to denote the two failure cases (None and StopEarly) for this?